File: Solutions/Python/maximum_sum_circular_subarray.py
"""
    [MEDIUM]
    918. Maximum Sum Circular Subarray

    Concepts:
    - array
    - kadane's algorithm

    Stats:
        Runtime | 92 ms     [Beats 77.21%]
        Memory  | 21.34 MB  [Beats 50.59%]
"""

class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        curr_max, curr_min = 0, 0
        max_sum, min_sum = -float("inf"), float("inf")

        for num in nums:
            curr_max = max(curr_max + num, num)
            curr_min = min(curr_min + num, num)
            max_sum = max(curr_max, max_sum)
            min_sum = min(curr_min, min_sum)

        return (
            max(max_sum, sum(nums) - min_sum)
            if max_sum > 0
            else max_sum
        )

File: Solutions/Cpp/maximum_sum_circular_subarray.cpp
/*
    [MEDIUM]
    918. Maximum Sum Circular Subarray

    Concepts:
    - array
    - kadane's algorithm

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 43.68 MB  [Beats 49.60%]
*/

class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int curr_max = 0, curr_min = 0;
        int max_sum = std::numeric_limits<int>::min();
        int min_sum = std::numeric_limits<int>::max();

        for (const int num : nums) {
            curr_max = std::max(curr_max + num, num);
            curr_min = std::min(curr_min + num, num);
            max_sum = std::max(curr_max, max_sum);
            min_sum = std::min(curr_min, min_sum);
        }

        return (max_sum > 0) ?
            (
                std::max(
                    max_sum,
                    std::accumulate(nums.begin(), nums.end(), 0) - min_sum
                )
            ) :
            (max_sum);
    }
};